### PR TITLE
Added oracle-based scenario runner

### PR DESCRIPTION
**Pull request by sifraser**

### CODE REVIEW

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            generatedScenario = GeneratedScenario(generatedScenario, actionCreators)
+            generateTaskResult
+        }
+        val queryTaskResult = query(testInstance, generatedScenario)
+        val answerCount = queryTaskResult.result?.size ?: 0
+        Log.debug("Query result, oracle=$oracleInstance answerCount=$answerCount")
+        if (answerCount == 0) {
+            return OracleResult(false,
+                oracleInstance.instanceName,
+                listOf(declarationValidTaskResult, buildActionsValidTaskResult) + generateTaskResults + queryTaskResult
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:72]: Why add `generateTaskResults` and `queryTaskResult` separately?
```suggestion
                listOf(declarationValidTaskResult, buildActionsValidTaskResult, generateTaskResults, queryTaskResult)
```

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            )
+        }
+        val verifyTaskResult = answersSufficient(testInstance, generatedScenario, queryTaskResult.result!!)
+        return OracleResult(
+            verifyTaskResult.result == true,
+            oracleInstance.instanceName,
+            listOf(declarationValidTaskResult,
+                buildActionsValidTaskResult) + generateTaskResults + queryTaskResult + verifyTaskResult
```
Comment (donbex) [inline azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:80]: Same.

**Review by donbex — CHANGES_REQUESTED** [review]

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/ImplementationInstance.kt [hunk 3]
```diff
@@ -149,22 +152,22 @@ class JarImplementationInstance<
...
         thread.start()
         thread.join()
         return runner.result!!
     }
 
-    private fun getClassloader(parent: ClassLoader): ClassLoader {
+    private fun getClassLoader(parent: ClassLoader): ClassLoader {
         if (!classLoaderCache.containsKey(parent)) {
-            Log.debug("Cache miss for $jar, parent class loader $parent")
+            Log.debug("Cache miss instance=$this jar=$jar, parentClasLoader=$parent")
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/ImplementationInstance.kt:175]: other way around .. and I'm missing an s :)

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Query.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+object NoQueryFactory : QueryFactory
+
+interface Query<T> : ReifiedBehavior
+
+interface DerivedQuery<T>
+
+class UnsupportedQuery<T> : Query<T> {
+    override val behavior = unsupportedBehavior
+}
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Query.kt:13]: how can it be?

[the remark this answers is not in the document]
Comment (donbex) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Query.kt:13]: In the same way that [UnsupportedCheck](https://github.com/anaplan-engineering/azuki/blob/2329c3733c89978c8b1923052e80d60794f52da6/azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Check.kt#L11-L13) is... unless the type parameter forces it to be a class?

File: azuki-runner/src/main/kotlin/com/anaplan/engineering/azuki/runner/ScenarioScriptRunner.kt [hunk 5]
```diff
@@ -83,6 +101,7 @@ fun main(args: Array<String>) {
     val script = File(scriptFile).readText()
     val imports = File(importsFile).readText()
     val oracleImpls = args.drop(3)
+    ScenarioScriptRunner.Log.debug("Starting script runner: testImpl=$testImpl oracleImpls=$oracleImpls")
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-runner/src/main/kotlin/com/anaplan/engineering/azuki/runner/ScenarioScriptRunner.kt:104]: we don't use commas for these

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+class MultiOracleScenarioRunner<
+    AF : ActionFactory,
+    CF : CheckFactory,
+    QF : QueryFactory,
+    AGF : ActionGeneratorFactory,
+    >(
+    private val testInstance: ImplementationInstance<AF, CF, QF, AGF>,
+    private val oracleInstances: List<ImplementationInstance<AF, CF, QF, AGF>>,
+    private val scenario: OracleScenario<AF, QF, AGF>,
+    private val runName: String,
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:16]: future enhancements :)

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+            oracleInstance.instanceName,
+            listOf(declarationValidTaskResult,
+                buildActionsValidTaskResult) + generateTaskResults + queryTaskResult + verifyTaskResult
+        )
+    }
+
+
+    private fun isDeclarationValid(instance: ImplementationInstance<AF, CF, QF, AGF>) =
+        instance.runTask("check-declaration") { implementation ->
+            val systemFactory = implementation.createSystemFactory()
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:87]: no

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+        }
+        val buildActionsValidTaskResult = areBuildActionsValid(oracleInstance)
+        if (buildActionsValidTaskResult.result != true) {
+            return OracleResult(false,
+                oracleInstance.instanceName,
+                listOf(declarationValidTaskResult, buildActionsValidTaskResult))
+        }
+        // TODO - c/should we cache generate/query of test system
+        var generatedScenario: OracleScenario<AF, QF, AGF> = scenario
+        val generateTaskResults = actionGenerators.map { generators ->
```
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:60]: I need both the list of results and the mutated scenario; I'm sure I could mash them together into a contrived fold, but to me that's just dogmatic

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt (hunk 1 above)
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:72]: maybe that's why you suggest a fold; your suggestion is not permitted `generateTaskResults` is plural

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt (hunk 2 above)
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:80]: and same comment

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/runner/MultiOracleScenarioRunner.kt:90]: p.s. remember most runs are executed on dedicated jvm

File: azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Query.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (sifraser) [reply azuki-core/src/main/kotlin/com/anaplan/engineering/azuki/core/system/Query.kt:13]: indeed, you can't have a polymorphic object -- it doesn't make any sense as it's by definition a singleton and it's nature cannot vary

**Pull request merged**